lib/util/node-event-generator.js

```diff
@@ -38,6 +49,16 @@ NodeEventGenerator.prototype = {
      */
     enterNode: function enterNode(node) {
         this.emitter.emit(node.type, node);
+
+        if (node.parent) {
+            this.currentAncestry.unshift(node.parent);
+        }
+
+        this.parsedQueryMap.forEach((parsedQuery, rawQuery) => {
+            if (esquery.matches(node, parsedQuery, this.currentAncestry)) {
```

Comment:
This is where I think we are going to have performance impact on the large ASTs. `esquery.matches` will traverse all ancestors to the top to check if the current node matches the selector. I know my original WIP PR did the same thing, but in my case, I wanted to later enhance it to read selectors from left to right and mark nodes as found as traversing occurred (and reset them on `:exit`). That would optimize performance and would have minimal impact on overall speed. In this case, I don't think we will have that option (unless we want to tweak esquery itself, which is an option).